lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs

```diff
@@ -103,135 +73,47 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask, IObserver<IClo
-                    _cancellationSource.IsCancellationRequested);
-            }
-            catch (Exception e)
-            {
-                if (e is IOException || e is TcpClientConnectionException || e is RemotingException ||
-                    e is SocketException)
-                {
-                    Logger.Log(Level.Error,
-                        "Received Exception {0} in MapTaskHost with message: {1}. The cancellation token is: {2}.",
-                        e.GetType(),
-                        e.Message,
-                        _cancellationSource.IsCancellationRequested);
-                    if (!_cancellationSource.IsCancellationRequested)
-                    {
-                        Logger.Log(Level.Error,
-                            "MapTask is throwing IMRUTaskGroupCommunicationException with cancellation token: {0}.",
-                            _cancellationSource.IsCancellationRequested);
-                        throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
-                    }
-                }
-                else if (e is AggregateException)
+
+                using (
+                    MapInputWithControlMessage<TMapInput> mapInput =
+                        _dataAndMessageReceiver.Receive(_cancellationSource))
                 {
-                    Logger.Log(Level.Error,
-                        "Received AggregateException. The cancellation token is: {0}.",
-                        _cancellationSource.IsCancellationRequested);
-                    if (e.InnerException != null)
-                    {
-                        Logger.Log(Level.Error,
-                            "InnerException {0}, with message {1}.",
-                            e.InnerException.GetType(),
-                            e.InnerException.Message);
-                    }
-                    if (!_cancellationSource.IsCancellationRequested)
+                    controlMessage = mapInput.ControlMessage;
+                    if (controlMessage != MapControlMessage.Stop)
                     {
-                        if (e.InnerException != null && e.InnerException is IOException)
+                        TMapOutput output = default(TMapOutput);
+                        try
                         {
-                            Logger.Log(Level.Error,
-                                "MapTask is throwing IMRUTaskGroupCommunicationException with cancellation token: {0}.",
-                                _cancellationSource.IsCancellationRequested);
-                            throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
+                            output = _mapTask.Map(mapInput.Message);
                         }
-                        else
+                        catch (Exception e)
                         {
-                            throw e;
+                            HandleTaskAppException(e);
```

Comment:
From examples, I can see it is more about calculations. Communication with other nodes to get data should happen before calling the Map() method.